### PR TITLE
feat(operator): Add scope params for step startedAt and finishedAt #3708

### PR DESCRIPTION
**Pull request by sh-tatsuno**
enhancement discussed in https://github.com/argoproj/argo/issues/3708

Checklist:

* [x] (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community
* [x] The title of the PR is (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed the CLA.
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My builds are green. Try syncing with master if they are not.

### CODE REVIEW

File: workflow/controller/operator.go [hunk 1]
```diff
@@ -2161,6 +2161,20 @@ func (woc *wfOperationCtx) buildLocalScope(scope *wfScope, prefix string, node *
...
 	}
 
+	if node.ID != "" {
+		key := fmt.Sprintf("%s.id", prefix)
+		scope.addParamToScope(key, node.ID)
+	}
+
+	if !node.StartedAt.Time.IsZero() {
+		key := fmt.Sprintf("%s.startedAt", prefix)
+		scope.addParamToScope(key, node.StartedAt.Time.Format("Mon Jan _2 15:04:05 2006"))
```
Comment (alexec) [inline workflow/controller/operator.go:2171]: should this be RFC3339?

Comment (sh-tatsuno) [reply workflow/controller/operator.go]: Thx! I 've fixed it!

File: docs/variables.md [hunk 2]
```diff
@@ -23,9 +26,12 @@ The following variables are made available to reference various metadata of a wo
 ## DAG Templates
 | Variable | Description|
 |----------|------------|
+| `steps.<STEPNAME>.id` | unique id of container task |
```
Comment (simster7) [inline docs/variables.md:29]:
```suggestion
| `tasks.<STEPNAME>.id` | unique id of container task |
```

File: workflow/controller/operator.go [hunk 3]
```diff
@@ -2161,6 +2161,20 @@ func (woc *wfOperationCtx) buildLocalScope(scope *wfScope, prefix string, node *
...
 
+	if node.ID != "" {
+		key := fmt.Sprintf("%s.id", prefix)
+		scope.addParamToScope(key, node.ID)
+	}
+
+	if !node.StartedAt.Time.IsZero() {
+		key := fmt.Sprintf("%s.startedAt", prefix)
+		scope.addParamToScope(key, node.StartedAt.Time.Format(time.RFC3339))
+	}
```
Comment (simster7) [inline workflow/controller/operator.go:2172]: Minor: space after clause
```suggestion
	}

```

File: docs/variables.md [hunk 4]
```diff
@@ -23,9 +26,12 @@ The following variables are made available to reference various metadata of a wo
...
 | `tasks.<TASKNAME>.ip` | IP address of a previous daemon container task |
 | `tasks.<TASKNAME>.status` | Phase status of any previous task |
 | `tasks.<TASKNAME>.exitCode` | Exit code of any previous script or container task |
+| `steps.<STEPNAME>.startedAt` | Timestamp when the task started |
+| `steps.<STEPNAME>.finishedAt` | Timestamp when the task finised |
```
Comment (simster7) [inline docs/variables.md:34]: Same here

File: docs/variables.md (hunk 2 above)
Comment (sh-tatsuno) [reply docs/variables.md]: sorry, they are quite careless mistakes.

File: workflow/controller/operator.go (hunk 3 above)
Comment (sh-tatsuno) [reply workflow/controller/operator.go:2172]: fixed to insert new line

File: docs/variables.md [hunk 5]
```diff
@@ -23,11 +26,14 @@ The following variables are made available to reference various metadata of a wo
...
 |----------|------------|
+| `tasks.<TASKNAME>.id` | unique id of container task |
 | `tasks.<TASKNAME>.ip` | IP address of a previous daemon container task |
 | `tasks.<TASKNAME>.status` | Phase status of any previous task |
 | `tasks.<TASKNAME>.exitCode` | Exit code of any previous script or container task |
+| `tasks.<TASKNAME>.startedAt` | Timestamp when the task started |
+| `tasks.<TASKNAME>.finishedAt` | Timestamp when the task finished |
 | `tasks.<TASKNAME>.outputs.result` | Output result of any previous container or script task |
-| `tasks.<STEPNAME>.outputs.parameters` | When the previous task uses 'withItems', this contains a JSON array of the output parameters of each invocation |
+| `tasks.<TASKNAME>.outputs.parameters` | When the previous task uses 'withItems', this contains a JSON array of the output parameters of each invocation |
```
Comment (simster7) [inline docs/variables.md:36]: Thanks for this fix!

**Pull request merged**